eureka-client/src/main/java/com/netflix/discovery/DiscoveryManager.java

```diff
@@ -40,31 +46,80 @@
  *
  */
 @Deprecated
+@Singleton
 public class DiscoveryManager {
     private static final Logger logger = LoggerFactory.getLogger(DiscoveryManager.class);
-    private DiscoveryClient discoveryClient;
 
-    private EurekaInstanceConfig eurekaInstanceConfig;
-    private EurekaClientConfig eurekaClientConfig;
-    private static final DiscoveryManager s_instance = new DiscoveryManager();
+    private static volatile DiscoveryManager INSTANCE;
+
+    @Inject
+    private volatile Provider<EurekaInstanceConfig> eurekaInstanceConfig = new Provider<EurekaInstanceConfig>() {
+        @Override
+        public EurekaInstanceConfig get() {
+            return null;
+        }
+    };
+
+    @Inject
+    private volatile Provider<EurekaClientConfig> eurekaClientConfig = new Provider<EurekaClientConfig>() {
+        @Override
+        public EurekaClientConfig get() {
+            return null;
+        }
+    };
+
+    @Inject
+    private volatile Provider<EurekaClient> eurekaClient = new Provider<EurekaClient>() {
+        @Override
+        public EurekaClient get() {
+            return null;
+        }
+    };
 
+    @Inject
     private DiscoveryManager() {
+        // static singleton hack
+        INSTANCE = this;
     }
 
-    public static DiscoveryManager getInstance() {
-        return s_instance;
+    public static synchronized DiscoveryManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new DiscoveryManager();
+        }
+        return INSTANCE;
     }
 
-    public void setDiscoveryClient(DiscoveryClient discoveryClient) {
-        this.discoveryClient = discoveryClient;
+    public synchronized void setDiscoveryClient(final DiscoveryClient client) {
+        if (this.eurekaClient.get() == null) {
+            this.eurekaClient = new Provider<EurekaClient>() {
+                @Override
+                public EurekaClient get() {
+                    return client;
+                }
+            };
+        }
     }
 
-    public void setEurekaClientConfig(EurekaClientConfig eurekaClientConfig) {
-        this.eurekaClientConfig = eurekaClientConfig;
+    public synchronized void setEurekaClientConfig(final EurekaClientConfig config) {
+        if (this.eurekaClientConfig.get() == null) {
+            this.eurekaClientConfig = new Provider<EurekaClientConfig>() {
+                @Override
+                public EurekaClientConfig get() {
+                    return config;
+                }
+            };
+        }
     }
 
-    public void setEurekaInstanceConfig(EurekaInstanceConfig eurekaInstanceConfig) {
-        this.eurekaInstanceConfig = eurekaInstanceConfig;
+    public synchronized void setEurekaInstanceConfig(final EurekaInstanceConfig config) {
```

Comment:
My point is that you don't need to write any code here.  Annotating this method with @Inject can be done instead of annotating the 'eurekaInstanceConfig' field with @Inject.